src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts

```diff
@@ -1127,29 +1126,29 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 				windowsShellArgs = true;
 				// If we don't have a cwd, then the terminal uses the home dir.
 				const userHome = await this._pathService.userHome();
-				if (basename === 'cmd.exe' && ((options.cwd && isUNC(options.cwd)) || (!options.cwd && isUNC(userHome.fsPath)))) {
-					return undefined;
-				}
-				if ((basename === 'powershell.exe') || (basename === 'pwsh.exe')) {
-					if (!shellSpecified) {
+				if (basename === 'cmd.exe') {
+					if ((options.cwd && isUNC(options.cwd)) || (!options.cwd && isUNC(userHome.fsPath))) {
+						return undefined;
+					}
+					if (!shellArgsSpecified) {
+						toAdd.push('/d', '/c');
+					}
+				} else if ((basename === 'powershell.exe') || (basename === 'pwsh.exe')) {
+					if (!shellArgsSpecified) {
 						toAdd.push('-Command');
 					}
 				} else if ((basename === 'bash.exe') || (basename === 'zsh.exe')) {
 					windowsShellArgs = false;
-					if (!shellSpecified) {
+					if (!shellArgsSpecified) {
```

Comment:
I think that we still want to also check if `!shellSpecified`, @Legend-Master?

Reply:
Do you mean we shouldn't add `-c` in this case? since you said yes in https://github.com/microsoft/vscode/pull/181760#discussion_r1301181334 ?

> ```jsonc
> "shell": {
>     "executable": "C:/Program Files/Git/bin/bash.exe",
> }
> ```

---

I think the right behavior is

- Add specific arguments for known shells
- Add `-c` for other shells
- Add nothing when the user sets `args` in the settings

**Add `/d /c` for this**

```jsonc
"shell": {
    "executable": "cmd.exe",
}
```

**Add `-c` for these**

```jsonc
"shell": {
    "executable": "C:/Program Files/Git/bin/bash.exe",
}
```
```jsonc
"shell": {
    "executable": "some_random_shell.exe",
}
```

**Do nothing for these**

```jsonc
"shell": {
    "executable": "some_random_shell.exe",
    "args": ["-whatever"],
}
```
```jsonc
"shell": {
    "executable": "some_random_shell.exe",
    "args": [],
}
```

---

In this way, we'll have the most out of box experience while still having full control over the arguments

What do you think?